### ISSUE TITLE
UndoRedo crashes godot in specific situations 

### ISSUE THREAD

**Issue by qarmin** [godotengine/godot#50769]
### Godot version

v3.4.beta.custom_build [40b57319e]

### System information

Ubuntu 21.04, Intel HD 3000

### Issue description

When executing UndoRedo functions sometimes Godot crash. 
This may be a problem with user logic, but I think that this should be handled at least in debug builds, because it cause a lot of crashes in Godot(#48796, #41737, #37385) and other projects (https://github.com/Orama-Interactive/Pixelorama/issues/421).

After executing 
```
	var temp_variable = ClassDB.instance("UndoRedo")
	temp_variable.create_action(".", -16)
	temp_variable.add_do_reference(SpatialMaterial.new())
	temp_variable.commit_action()
	temp_variable.undo()
	temp_variable.create_action(".", -42)
```
cause this 
```
==33941==ERROR: AddressSanitizer: heap-use-after-free on address 0x61900017ca90 at pc 0x000001864727 bp 0x7ffd93e7a920 sp 0x7ffd93e7a910
READ of size 8 at 0x61900017ca90 thread T0
    #0 0x1864726 in Ref<Reference>::unref() core/reference.h:257
    #1 0x18546b8 in Ref<Reference>::~Ref() core/reference.h:272
    #2 0x11d252b3 in UndoRedo::Operation::~Operation() core/undo_redo.h:55
    #3 0x11d31b93 in List<UndoRedo::Operation, DefaultAllocator>::Element::~Element() core/list.h:51
    #4 0x11d31bed in void memdelete_allocator<List<UndoRedo::Operation, DefaultAllocator>::Element, DefaultAllocator>(List<UndoRedo::Operation, DefaultAllocator>::Element*) core/os/memory.h:127
    #5 0x11d2f052 in List<UndoRedo::Operation, DefaultAllocator>::_Data::erase(List<UndoRedo::Operation, DefaultAllocator>::Element const*) core/list.h:168
    #6 0x11d29e33 in List<UndoRedo::Operation, DefaultAllocator>::erase(List<UndoRedo::Operation, DefaultAllocator>::Element const*) core/list.h:356
    #7 0x11d2a048 in List<UndoRedo::Operation, DefaultAllocator>::clear() core/list.h:389
    #8 0x11d25b8f in List<UndoRedo::Operation, DefaultAllocator>::~List() core/list.h:692
    #9 0x11d249a0 in UndoRedo::Action::~Action() core/undo_redo.h:69
    #10 0x11d2e622 in CowData<UndoRedo::Action>::_unref(void*) core/cowdata.h:211
    #11 0x11d28b74 in CowData<UndoRedo::Action>::resize(int) core/cowdata.h:269
    #12 0x11d25874 in Vector<UndoRedo::Action>::resize(int) core/vector.h:86
    #13 0x11d0c14e in UndoRedo::_discard_redo() core/undo_redo.cpp:53
    #14 0x11d0c5a8 in UndoRedo::create_action(String const&, UndoRedo::MergeMode) core/undo_redo.cpp:60
    #15 0x11d393bc in MethodBind2<String const&, UndoRedo::MergeMode>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:1523
    #16 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #17 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #18 0x1e2bfb1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1034
    #19 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #20 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #21 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #22 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #23 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #24 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #25 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #26 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #27 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #28 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #29 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #30 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #31 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #32 0x7fd914c12564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)
    #33 0x17a79ad in _start (/usr/bin/godots+0x17a79ad)

0x61900017ca90 is located 16 bytes inside of 928-byte region [0x61900017ca80,0x61900017ce20)
freed by thread T0 here:
    #0 0x7fd915b7a8f7 in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:127
    #1 0x120b0cf2 in Memory::free_static(void*, bool) core/os/memory.cpp:168
    #2 0x1969f91 in void memdelete<Object>(Object*) core/os/memory.h:118
    #3 0x11d0bee9 in UndoRedo::_discard_redo() core/undo_redo.cpp:46
    #4 0x11d0c5a8 in UndoRedo::create_action(String const&, UndoRedo::MergeMode) core/undo_redo.cpp:60
    #5 0x11d393bc in MethodBind2<String const&, UndoRedo::MergeMode>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:1523
    #6 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #7 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #8 0x1e2bfb1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1034
    #9 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #10 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #11 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #12 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #13 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #14 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #15 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #16 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #17 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #18 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #19 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #20 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #21 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #22 0x7fd914c12564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)

previously allocated by thread T0 here:
    #0 0x7fd915b7ac47 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x120afcb3 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:75
    #2 0x120afbc4 in operator new(unsigned long, char const*) core/os/memory.cpp:40
    #3 0xc317ab7 in Object* ClassDB::creator<SpatialMaterial>() core/class_db.h:140
    #4 0x118e44b6 in ClassDB::instance(StringName const&) core/class_db.cpp:520
    #5 0x1c37940 in GDScriptNativeClass::instance() modules/gdscript/gdscript.cpp:77
    #6 0x1c37328 in GDScriptNativeClass::_new() modules/gdscript/gdscript.cpp:65
    #7 0x1d12da7 in MethodBind0R<Variant>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:237
    #8 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #9 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #10 0x1e2bf30 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1032
    #11 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #12 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #13 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #14 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #15 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #16 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #17 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #18 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #19 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #20 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #21 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #22 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #23 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #24 0x7fd914c12564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)

SUMMARY: AddressSanitizer: heap-use-after-free core/reference.h:257 in Ref<Reference>::unref()

```

or 
```
	var temp_variable = ClassDB.instance("UndoRedo")
	temp_variable.create_action(".", -19)
	temp_variable.add_do_reference(AudioEffectEQ21.new())
	temp_variable.clear_history(true)
	temp_variable.commit_action()
	temp_variable.create_action(".", 1)
```
```
ERROR: Condition "action_level > 0" is true.
   at: clear_history (core/undo_redo.cpp:351)
=================================================================
==36657==ERROR: AddressSanitizer: heap-use-after-free on address 0x6150000af590 at pc 0x000001864727 bp 0x7fff5b7e3cf0 sp 0x7fff5b7e3ce0
READ of size 8 at 0x6150000af590 thread T0
    #0 0x1864726 in Ref<Reference>::unref() core/reference.h:257
    #1 0x18546b8 in Ref<Reference>::~Ref() core/reference.h:272
    #2 0x11d252b3 in UndoRedo::Operation::~Operation() core/undo_redo.h:55
    #3 0x11d31b93 in List<UndoRedo::Operation, DefaultAllocator>::Element::~Element() core/list.h:51
    #4 0x11d31bed in void memdelete_allocator<List<UndoRedo::Operation, DefaultAllocator>::Element, DefaultAllocator>(List<UndoRedo::Operation, DefaultAllocator>::Element*) core/os/memory.h:127
    #5 0x11d2f052 in List<UndoRedo::Operation, DefaultAllocator>::_Data::erase(List<UndoRedo::Operation, DefaultAllocator>::Element const*) core/list.h:168
    #6 0x11d29e33 in List<UndoRedo::Operation, DefaultAllocator>::erase(List<UndoRedo::Operation, DefaultAllocator>::Element const*) core/list.h:356
    #7 0x11d25ae1 in List<UndoRedo::Operation, DefaultAllocator>::pop_front() core/list.h:279
    #8 0x11d0d29e in UndoRedo::create_action(String const&, UndoRedo::MergeMode) core/undo_redo.cpp:80
    #9 0x11d393bc in MethodBind2<String const&, UndoRedo::MergeMode>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:1523
    #10 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #11 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #12 0x1e2bfb1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1034
    #13 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #14 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #15 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #16 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #17 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #18 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #19 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #20 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #21 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #22 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #23 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #24 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #25 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #26 0x7fddd54f0564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)
    #27 0x17a79ad in _start (/usr/bin/godots+0x17a79ad)

0x6150000af590 is located 16 bytes inside of 504-byte region [0x6150000af580,0x6150000af778)
freed by thread T0 here:
    #0 0x7fddd64588f7 in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:127
    #1 0x120b0cf2 in Memory::free_static(void*, bool) core/os/memory.cpp:168
    #2 0x1969f91 in void memdelete<Object>(Object*) core/os/memory.h:118
    #3 0x11d0cfce in UndoRedo::create_action(String const&, UndoRedo::MergeMode) core/undo_redo.cpp:75
    #4 0x11d393bc in MethodBind2<String const&, UndoRedo::MergeMode>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:1523
    #5 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #6 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #7 0x1e2bfb1 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1034
    #8 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #9 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #10 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #11 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #12 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #13 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #14 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #15 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #16 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #17 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #18 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #19 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #20 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #21 0x7fddd54f0564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)

previously allocated by thread T0 here:
    #0 0x7fddd6458c47 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x120afcb3 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:75
    #2 0x120afbc4 in operator new(unsigned long, char const*) core/os/memory.cpp:40
    #3 0x10104124 in Object* ClassDB::creator<AudioEffectEQ21>() core/class_db.h:140
    #4 0x118e44b6 in ClassDB::instance(StringName const&) core/class_db.cpp:520
    #5 0x1c37940 in GDScriptNativeClass::instance() modules/gdscript/gdscript.cpp:77
    #6 0x1c37328 in GDScriptNativeClass::_new() modules/gdscript/gdscript.cpp:65
    #7 0x1d12da7 in MethodBind0R<Variant>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:237
    #8 0x11b6557b in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #9 0x11ded653 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1173
    #10 0x1e2bf30 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1032
    #11 0x1c5d2e3 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1181
    #12 0x1c5d596 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1187
    #13 0xc3b2c3e in Node::_notification(int) scene/main/node.cpp:146
    #14 0x1b08395 in Node::_notificationv(int, bool) scene/main/node.h:45
    #15 0x1b0a80a in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:162
    #16 0xdf950b8 in Node2D::_notificationv(int, bool) scene/2d/node_2d.h:37
    #17 0x11b65a15 in Object::notification(int, bool) core/object.cpp:927
    #18 0xc3b4bbb in Node::_propagate_ready() scene/main/node.cpp:187
    #19 0xc3b4421 in Node::_propagate_ready() scene/main/node.cpp:179
    #20 0xc4063b4 in Node::_set_tree(SceneTree*) scene/main/node.cpp:2557
    #21 0xc4cc300 in SceneTree::init() scene/main/scene_tree.cpp:462
    #22 0x183c67d in OS_X11::run() platform/x11/os_x11.cpp:3626
    #23 0x17a7d8b in main platform/x11/godot_x11.cpp:55
    #24 0x7fddd54f0564 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28564)

SUMMARY: AddressSanitizer: heap-use-after-free core/reference.h:257 in Ref<Reference>::unref()
```

### Steps to reproduce

Above

**Comment by qarmin** [comment]
Reopening since only first example was fixed, second still use invalid memory

**Comment by akien-mga** [comment]
Fixed by #54433.

**Comment by nabfrew** [comment]
I think this should be re-opened as the fix was reverted?
